### ISSUE TITLE
GPU 500x faster than CPU!?

### ISSUE THREAD

**Issue by cmis91** [facebookresearch/pytorch3d#469]
I literally ran the camera_position_optimization_with_differentiable_rendering.ipynb twice using GPU and CPU respectively without modifying any codes from the tutorial 

Use a general Intel i7 CPU, 
Optimizing (loss 1303.0554): 10%           20/200 [02:51<27:08, **9.05s/it**]
loss.backward() take on average about **7 seconds** and forward takes around 0.8 seconds

Use my destop GTX 1080Ti
Optimizing (loss 531.0090): 100%            200/200 [00:04<00:00, **40.98it/s**]
forward time: 0.005010562999814283 seconds
backward time: **0.09047832899977948 seconds**

So, is the GPU  500x faster than CPU? I would expect ~10x faster if switching from CPU to GPU. Can anyone who really knows what's happening under the hood help explain why GPU is so much faster, for this particular case, and some possible direction where we can optimize the CPU run time? 

Thanks a lot.

**Comment by jcjohnson** [comment]
This is not very surprising. Our CPU rasterizers are very naive and not well-optimized at all, since in practice we expect many more people to train on GPU vs CPU. For CPU mesh rendering, I expect the forward rasterizer to be the bottleneck:

https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/csrc/rasterize_meshes/rasterize_meshes_cpu.cpp#L107

It does a brute-force comparison of every pixel with every face in the mesh, and in particular does *not* use the same tiled coarse-to-fine approach that we use in the forward pass of the GPU rasterizer. One bit of easy low-hanging fruit for improving the CPU rasterizer would be to implement the same coarse-to-fine logic as we use on the CPU -- we actually already have the first step (coarse rasterization) implemented on CPU:

https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/csrc/rasterize_meshes/rasterize_meshes_cpu.cpp#L414

so you'd just need to implement a CPU function for the second step (fine rasterization) and plug it in here:

https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/csrc/rasterize_meshes/rasterize_meshes.h#L390

Is CPU runtime actually a concern for you, or are you more just curious about what's going on?

**Comment by cmis91** [comment]
Thank you for the insightful response. It's all about the cost. The cheapest AWS GPU instance costs $0.5/hour and the CPU instance with a similar setup costs a fraction of that. Our application is to return a corresponding 3D model from a 3D model database given an input reference image. Running such an application 24/7 on GPU instances is pricy.

Slightly off the topic. For pose estimation, currently, we implemented PWP3D http://www.robots.ox.ac.uk/~victor/pdfs/prisacariu_reid_ijcv2012_draft.pdf  with rendering in VTK. They use the posterior probability of each pixel being foreground or background as the loss (they call it energy) and apply simple gradient descent to optimize. Working with quaternions, they have a closed-form gradient. Do you think this will perform better or faster than the Autograd in Pytorch3D? 

Another question is about the local optimal. I did observe that both PWP3D and Pytorch suffer from the local optimal. For many initial poses, the optimization process could not convert to the reference position and stuck in the wrong pose. Do you think the end-to-end differentiable pipeline as in Pytorch3D can make it less prone to the starting pose position? 

The last question is that why you spend time implementing a rending function (rasterization) instead of using an existing optimized package (e.g., OpenGL, VTK).  Is it because other approaches are not differentiable? Sorry, didn't read your paper in great detail. 

Thanks in advance.

**Comment by jcjohnson** [comment]
Some of these are really more like research questions and not really issues with the library, but:

> Slightly off the topic. For pose estimation, currently, we implemented PWP3D http://www.robots.ox.ac.uk/~victor/pdfs/prisacariu_reid_ijcv2012_draft.pdf with rendering in VTK. They use the posterior probability of each pixel being foreground or background as the loss (they call it energy) and apply simple gradient descent to optimize. Working with quaternions, they have a closed-form gradient. Do you think this will perform better or faster than the Autograd in Pytorch3D?

I'm not familiar with PWP3D, so I can't really speak to the tradeoffs.

> Another question is about the local optimal. I did observe that both PWP3D and Pytorch suffer from the local optimal. For many initial poses, the optimization process could not convert to the reference position and stuck in the wrong pose. Do you think the end-to-end differentiable pipeline as in Pytorch3D can make it less prone to the starting pose position?

I think using gradient descent to optimize pose is going to be susceptible to local optima no matter what you do. The space of 3D poses over which you are optimizing is fundamentally low-dimensional, and local optima exist -- this is bad news for any vanilla gradient descent method. Some workarounds might be using a fancier optimizer (conjugate gradient, some second order method like L-BFGS), or possibly just trying multiple random restarts with gradient descent and taking the best result.

> The last question is that why you spend time implementing a rending function (rasterization) instead of using an existing optimized package (e.g., OpenGL, VTK). Is it because other approaches are not differentiable? Sorry, didn't read your paper in great detail.

I'm not familiar with VTK (but I suspect it probably uses something like OpenGL or DirectX under the hood). Existing low-level APIs like OpenGL are not differentiable out of the box. There are some differentiable renderers like [OpenDR](https://github.com/mattloper/opendr/wiki) or [Nvdiffrast](https://github.com/NVlabs/nvdiffrast) which use OpenGL in the forward pass, and implement custom code for the backward pass. This approach is indeed much faster than our software rasterizer (since OpenGL can make use of hardware-accelerated rasterization on the GPU) but it constrains you to use the precise rendering algorithms supported by OpenGL; we wanted the flexibility to implement different types of rendering algorithms, so we chose to implement our own.

**Comment by cmis91** [comment]
Thank you for the detailed explanation. Justin, I am glad to see your continuing involvement and support in this library given your busy academic and teaching schedule at U of Michigan. Close this question now. #Thanks to @jcjohnson
